### ISSUE TITLE
Clang invokes explicit conversion operator when initializing const ref of target type

### ISSUE THREAD

**Issue by muggenhor** [llvm/llvm-project#68123]
[Example](https://godbolt.org/z/Gsox1fhjx):
```cpp
struct T {};
struct U { explicit operator T() const; };

void f(const U& x) {
  const T& y(x); // <-- this is *not* an initialization of `T` but const-ref to `T`
}
```

I expect only *implicit* conversions to be allowed for intialization of `y` there. And GCC agrees (see godbolt link). Clang accepts this though.

I.e. out of these options it should choose the last, but uses the first instead:

1. convert to `T` with life-time extension as if by `const T& y = T(x);`
2. `reinterpret_cast` as if by `const T& y = (const T&)x;`
3. program ill-formed: emit error

When declaring a converting constructor instead Clang also agrees (and is thus inconsistent), [example 2](https://godbolt.org/z/KcsxGvzoP):
```cpp
struct U {};
struct T {
  explicit T(const U&);
};

void f(const U& x) {
  const T& y(x); // <-- this is *not* an initialization of `T` but const-ref to `T`
}
```

TLDR: for initializing of local (const?) references clang considers explicit conversion operators (but not explicit constructors) while it shouldn't.

**Comment by muggenhor** [comment]
I stand corrected. It looks like GCC is in error and Clang is correct:
> Anyway, as for the OP's question, Clang is right; https://eel.is/c++draft/dcl.init.ref#5.3.2 governs this initialization, and https://eel.is/c++draft/over.match.ref#1 allows an explicit conversion function to be a candidate in this context.
